apps/prairielearn/elements/pl-integer-input/pl-integer-input.mustache

```diff
@@ -16,30 +16,55 @@
             <span class="input-group-text" id="pl-integer-input-{{uuid}}-label">{{{label}}}</span>
         </span>
         {{/label}}
-        <input name="{{name}}" type="text" class="form-control pl-integer-input-input {{^display_append_span}}rounded-right{{/display_append_span}}" size="{{size}}" {{^editable}}disabled{{/editable}} {{#raw_submitted_answer}}value="{{raw_submitted_answer}}"{{/raw_submitted_answer}} aria-describedby="pl-integer-input-{{uuid}}-label pl-integer-input-{{uuid}}-suffix" {{#show_placeholder}}placeholder="{{shortinfo}}"{{/show_placeholder}} />
+        <input
+            name="{{name}}"
+            type="text"
```

Comment:
I'm not saying you should do it, I'm just raising a discussion.

Reply:
I think most of the time this will be an integer in base 10, but we could add a check for this if necessary.

Reply:
Not relevant for accessibility, but it is a nice UX enhancement for folks working from a mobile device. I'm in favor of `inputmode="numeric"` here, of course with a check for an appropriate base.

Reply:
I think a combination of inputmode with a [`pattern="[0-9a-fA-F_]+"`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern) for base 16 (or equivalents in other bases or for special cases like allow-blank and base=0) would be a great addition.

Reply:
Tried doing this here: https://github.com/PrairieLearn/PrairieLearn/pull/6980/commits/bba0fca7d61e325bed2f7f28bba7bb9b9b78e919

Reply:
@jonatanschroeder I'm actually not totally sure if the pattern is a good idea. It prevents submission of invalid characters, but I think we actually want to allow student submissions in that case and then just show the invalid message. What are your thoughts on this?

Reply:
I think the type can remain `text` (which would allow text to get an invalid input), only `inputmode` (and possibly `pattern`) can be set. In that case, students will have the option of writing something invalid and still get feedback. The input mode is mostly a way to allow students to see a suitable keyboard in onscreen keyboard environments (e.g., mobile, tablets).

The pattern is a way to provide client-side validation, but not sure if and how we want to handle that (a warning, [maybe via Bootstrap classes?](https://getbootstrap.com/docs/4.6/components/forms/#validation) @nwalters512 thoughts?).

About the code: you use numeric for bases 1-9, it should be 1-10. As for the pattern (assuming we keep using it), I kinda liked your idea of using `string` constants. Just make sure your pattern allows `_` characters (plus whatever else is acceptable, spaces?). Also, use `*` instead of `+` if allow blank is set. About base 0, I'd suggest something like: `([0-9_]+|0[xX][0-9a-fA-F_]+|0[oO][0-7_]+|0[bB][01_]+)`, which would allow any valid combination of numbers acceptable in that base option.

Reply:
IIRC the browser itself will warn if input doesn't match the pattern, so we shouldn't require any custom CSS/markup for that (🤞).

Even if we don't use `pattern` (or don't want to decide that at the moment), `inputmode="numeric"` for appropriate bases is a great starting point.

Reply:
> IIRC the browser itself will warn if input doesn't match the pattern, so we shouldn't require any custom CSS/markup for that (🤞).

This is actually what I was concerned about. My browser wouldn't let me submit something (or at least, I wasn't sure how to dismiss the warning) that didn't match the pattern when I was playing with this. @nwalters512 I agree with your comment about allowing unconditional submission of invalid-ish stuff, so I think we should get rid of the pattern if this is the behavior.

Reply:
@nwalters512 updated here https://github.com/PrairieLearn/PrairieLearn/pull/6980/commits/41881e12afdd1b6c405a9d56ba8a9aaa28a144d3

Removed the pattern checking since this prevented submission of arbitrary text (which is the goal of always using `type="text"`).